accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,871 @@
+        element.
+
+*   An `ifElement` is potentially constant if its condition, "then" element, and
+    "else" element (if any) are potentially constant expressions.
+
+*   A `forElement` is never a constant element. A `for` element cannot be used
+    in constant collection literals.
+
+*   It is a compile-time error if any element in a constant set or key in a
+    constant map does not have a primitive operator `==`.
+
+*   It is a compile-time error if an element in a const set is equal to any
+    other element according to its operator `==`.
+
+*   It is a compile-time error if a key in a constant map is equal to any other
+    key according to its operator `==`.
+
+*   It is a compile-time error to have duplicate values in a set literal or
+    duplicate keys in a map literal, according to behavior specified in the
+    runtime semantics.
+
+*   When evaluating a const collection literal, if another const collection has
+    previously been evaluated that:
+
+    *   is of the same kind (map, set, or list),
+    *   has the same type arguments,
+    *   and contains the same (identical) elements or key/value entries in the
+        same order,
+
+    then the current literal evaluates to the value of that previous literal.
+    *In other words, constants are canonicalized.* Note that maps and sets are
+    only canonicalized if they contain the same keys or elements *in the same
+    order*.
+
+## Dynamic Semantics
+
+Spread, `if`, and `for` behave similarly across all collection types. To
+simplify the spec, there is a single procedure used for all kinds of
+collections. This recursive procedure builds up either a `result` sequence of
+values or key-value pair map entries. Then a final step handles those
+appropriately for the given collection type.
+
+### To evaluate a collection `element`:
+
+1.  If `element` is an expression element:
+
+    1.  Evaluate the element's expression and append it to `result`.
+
+1.  Else, `element` is a `mapEntry` `keyExpression: valueExpression`:
+
+    1.  Evaluate `keyExpression` to a value `key`.
+
+    1.  Evaluate `valueExpression` to a value `value`.
+
+    1.  Append `key: value` to `result`.
+
+1.  Else, if `element` is a spread element:
+
+    1.  Evaluate the spread expression to a value `spread`.
+
+    1.  If `entry` is null-aware and `spread` is null, do nothing.
+
+    1.  Otherwise:
+
+        1.  If the collection is a map, evaluate `spread.entries.iterator` to a
+            value `iterator`. Otherwise, evaluate `spread.iterator` to
+            `iterator`. *This will deliberately throw an exception if `spread`
+            is `null` and `element` is not null-aware.*
+
+        1.  Loop:
+
+            1.  If `iterator.moveNext()` returns `false`, exit the loop.
+
+            1.  Evaluate `iterator.current` and append it to `result`. *This
+                will be a MapEntry in a map literal, or any object for a list or
```

Comment:
You are adding a MapEntry to result here, but `key: value` above, which is not a `MapEntry`. You are also storing a user provided object until later, and it might be mutable.
I think here you should say:
> If collection is a map, evaluate `iterator.current` to a value *c*, then evaluate `c.key` to a value *k* and `c.value` to a value *v*. Then append *k*:*v* to *result*.
> Otherwise evaluate iterator.current to a value *e* and append *e* to *result*.

Reply:
Good catch. Went ahead and split out the map and non-map cases since they are fairly different now.